### DISCUSSION TITLE
Node.get_properties and similar methods

### DISCUSSION THREAD

**Discussion by tovaverm** [FreeOpcUa/opcua-asyncio#1357] [Q&A]
I have a Node ua_obj. I try to get the properties with ua_obj.get_properties() and then try to iterate over them, but I get the error
    TypeError: 'coroutine' object is not iterable
    sys:1: RuntimeWarning: coroutine 'Node.get_children' was never awaited

Further investigation leads me to the get_properties method of Node with call Node.get_children. This function is an async function but in get_properties is never awaited. get_properties is also not an async function.

I don't have a lot experience with asyncio but I believe this is a bug. The get_properties method should be async (as well as the get_variables and get_mthoedsfunction) and the get_children call should be awaited. 

Can this be confirmed?

**Comment by schroeder-** [comment]
Did you try this?

```Python
for child in await ua_obj.get_properties():
           .....
```

**Reply by tovaverm** [reply]
This works. Thanks for the suggestion. I wonder why it has been implemented this way. Why not do the await inside the get_properties function?

**Reply by schroeder-** [reply]
I don't know and it doesn't realy matter i think. Also the doc string in get_properties, hints that you get a coroutine back, which you then have to await.

**Reply by oroulet** [reply]
> Why not do the await inside the get_properties function?

What? that function is implemented as it should  with python asyncio. You cannot have a hidden await inside a function. you need to await it.

**Reply by tovaverm** [reply]
If you read the initial comment correctly, I suggest to make the get_properties function also a coroutine which can then be awaited.

**Comment by AndreasHeine** [comment]
> This works. Thanks for the suggestion. I wonder why it has been implemented this way. Why not do the await inside the get_properties function?

even if you are awaiting something inside of get_properties you still need to await the function itself (its asyncio not threading!)

**Reply by tovaverm** [reply]
If you read the initial post correctly, I suggest to make the get_properties function a coroutine which can be awaited. And inside the function, already await the get_children coroutine. But looking at it again, I notice the pattern that the get_properties is similar to a partial function. It returns the get_children coroutine with some arguments already fixed.
